src/npm/Prefetcher.ts

```diff
@@ -0,0 +1,84 @@
+import type { DocumentListParams } from 'nano';
+
+import { config } from '../config';
+import { log } from '../utils/log';
+import { wait } from '../utils/wait';
+
+import * as npm from './index';
+
+export type PrefetchedPkg = { id: string };
+
+export class Prefetcher {
+  #limit: number = config.bootstrapConcurrency;
+  #ready: PrefetchedPkg[] = [];
+  #nextKey: string | null = null;
+  #running: boolean = false;
+  #offset: number = 0;
+  #finished: boolean = false;
+  #maxIdle = 100;
+
+  constructor(opts: { nextKey: string | null }) {
+    this.#nextKey = opts.nextKey;
+  }
+
+  get offset(): number {
+    return this.#offset + this.#limit - this.#ready.length;
+  }
+
+  get idleCount(): number {
+    return this.#ready.length;
+  }
+
+  get isFinished(): boolean {
+    return this.#finished;
+  }
+
+  async getNext(): Promise<PrefetchedPkg> {
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      if (this.#ready.length <= 0) {
+        await wait(100);
+        continue;
+      }
+
+      return this.#ready.shift()!;
+    }
```

Comment:
```suggestion
    while (this.#ready.length <= 0) {
      await wait(100);
    }
    return this.#ready.shift()!;
```

should this not be the condition of the while, and the ready shift after it?

Reply:
good point, much cleaner 👌🏻